### ISSUE TITLE
Use inline cache to avoid the need to rebuild the Docker base layer

### ISSUE THREAD

**Issue by jameshalsall** [inviqa/harness-go#183]
Since we changed the Dockerfile structure to be a single, multi-stage build, we see that when building the production image (`ws build-prod`) it does not make use of the already-built base layer.

I believe this is possibly because we are building the base layer using `docker-compose build`, rather than `docker build`, and we may need to specify some additional options to get it to use the layer cache:

```
--build-arg BUILDKIT_INLINE_CACHE=1
--cache-from=<base-layer-tag>
```

**Comment by andytson-inviqa** [comment]
yep those can be put in the docker-bake.hcl, but for docker-compose.yml bake build I expect x-bake is needed, which is only currently supported on the linux-amd64/arm64 jenkins node labels.

I've wanted to try them out for a distributed cache, but sadly they don't cache buildkit `RUN --mount=type=cache...`, which is where I think the best performance comes in for php composer (since composer plugins love to hook in and change things), but `RUN --mount=type=cache` has more limited use in Go.

However due to your workflow of rebuilding go image every change for local dev, `RUN --mount=type=cache` would be worthwhile there keeping the GOPATH cached.

**Comment by andytson-inviqa** [comment]
https://docs.docker.com/compose/compose-file/compose-file-v3/#cache_from
https://docs.docker.com/compose/compose-file/compose-file-v3/#args

if not buildx

**Comment by andytson-inviqa** [comment]
I don't think this is the cause of your buildx way not using cache though, just I need to have some time to check what's happening

**Comment by jameshalsall** [comment]
To clarify, I am proposing switching the app build from `docker-compose build` to just `docker build`. The options in my original message are related to pure docker builds, the problem we have is that when building the prod layer we switch to `docker build` but we have not tagged the original base image layer, and cannot use it in the prod build with `--cache-from`. It may not be that simple, but according to https://pythonspeed.com/articles/faster-multi-stage-builds/ it should be possible to use the base image layer that has already been built, rather than building it again.

The local builds are not really an issue, Go modules only redownload if you edit your `go.mod|sum` file. However, even that won't matter eventually when we implement #161 fully.

**Comment by andytson-inviqa** [comment]
no need https://github.com/inviqa/harness-go/issues/183#issuecomment-976836892

**Comment by jameshalsall** [comment]
Yes, in `docker-compose`, but we don't use `docker-compose` to build the prod image.

**Comment by andytson-inviqa** [comment]
also one benefit of the prod image not having an inline cache is that the sensitive secrets in the base stage aren't uploaded. They can be changed to use docker build secrets, but I'd still like to see if there's a simpler solution
